### ISSUE TITLE
puzzled by empty coverage using package_coverage.

### ISSUE THREAD

**Issue by kforner** [jimhester/covr#188]
When I run package_coverage() on my package I get an empty coverage:
```
cov <- package_coverage(pkg$path, type = 'tests', use_subprocess = FALSE, use_try = TRUE, quiet = FALSE, clean = FALSE)
> length(cov)
[1] 0
```
looking at the generated files: 
1) my package is installed, and hooked:
```
cat qbutils/R/qbutils
...
local({
    info <- loadingNamespaceInfo()
    pkg <- info$pkgname
    ns <- .getNamespace(as.name(pkg))
    if (is.null(ns))
        stop("cannot find namespace environment for ", pkg, domain = NA);
    dbbase <- file.path(info$libname, pkg, "R", pkg)
    lazyLoad(dbbase, ns, filter = function(n) n != ".__NAMESPACE__.")
setHook(packageEvent(pkg, "onLoad"), function(...) covr:::trace_environment(ns))
reg.finalizer(ns, function(...) { covr:::save_trace("/tmp/Rtmpshbx1l/R_LIBS87d194643") }, onexit = TRUE)
})
```

2) the tests as executed by testInstalledPackage() seem to have run successfully:

```
cat qbutils-tests/testthat.Rout
| file         | F    | S | E | NOK | real |
|--------------+------+---+---+-----+------|
| string_utils | 0/98 | 0 |   |     | 0.48 |
====================== Overall Test Results ======================
| files | contexts | tests | F    | S | E | NOK | real | elapsed |
|-------+----------+-------+------+---+---+-----+------+---------|
| 1     | 0/1      | 0/10  | 0/98 | 0 |   |     | 0.48 | 1.41    |

RESULT: OK
> 
```
but there are no traces files in the directory. 

What should I look for ? 
thanks

**Comment by kforner** [comment]
I also tried with one test package from covr, with the same result:
```
/ (master) R -q
> library(devtools)
> load_all('.')
Loading covr
> cov=package_coverage('tests/testthat/TestUseTry/')
> length(cov)
[1] 0
```

I'm on commit c3efe845c93759c5b39e11d9e625df91bfa6bea9 , version 2.0.1.9000

**Comment by jimhester** [comment]
Have you installed the devel version of covr, not just using `load_all()`?

**Comment by kforner** [comment]
I have cloned the master, at commit c3efe845c93759c5b39e11d9e625df91bfa6bea9

**Comment by jimhester** [comment]
Yes but did you actually install it, or just load it in your current session using `devtools::load_all()`?

**Comment by kforner** [comment]
Sorry for the edit. You're right, I installed it but always loaded it via load_all. Let me try again. thanks.

**Comment by kforner** [comment]
loading via library(covr) solves the problem for 'tests/testthat/TestSummary', thank you.
But not for my package, I'll keep digging. 

Could you please explain why it is working when covr is installed and loaded via library, and not via load_all() ? 

Also, am I right by assuming that, if there are tests installed in a package, the covr trace files will be created at the end of the testInstalledPackage() execution ?

**Comment by jimhester** [comment]
> Could you please explain why it is working when covr is installed and loaded via library, and not via load_all() ?

`load_all()` only loads the package in the current session, but the tracing calls also need to call `covr::trace_environment()`, `covr:::count()` etc in the spawned subprocesses. So if you don't have covr installed these will fail. I have this same problem when doing covr development sometimes so I figured it might be the issue.

The tracing files are created by `covr:::save_trace()` when either the package namespace being tested is removed and garbage collected or the R session exits. see `?reg.finalizer`. So if there are tests the tracing files should be created after each subprocess exists. (1 per test file)

**Comment by kforner** [comment]
> the tracing calls also need to call covr::trace_environment(), covr:::count() etc in the spawned subprocesses. 
of course, now it seems so obvious: the spawned subprocesses find and use my old install of covr.

**Comment by kforner** [comment]
Ok, I finally figured it out. I use the R_LIBS var, which supersedes the R_LIBS_USER that you use (at https://github.com/jimhester/covr/blob/master/R/covr.R#L205), where my tested package can also be found.
Any reason not to use R_LIBS instead of R_LIBS_USER ?

**Comment by jimhester** [comment]
`tools::testInstalledPackage()` unsets `R_LIBS=` before running the tests.

[src/library/tools/R/testing.R#L315-L316](https://github.com/wch/r-source/blob/31af83a9fd8a0bf2e511cf9f59bebc251b845e18/src/library/tools/R/testing.R#L315-L316)

**Comment by kforner** [comment]
not when running tests:
https://github.com/wch/r-source/blob/31af83a9fd8a0bf2e511cf9f59bebc251b845e18/src/library/tools/R/testing.R#L407

**Comment by kforner** [comment]
and anyway you do NOT want to nullify R_LIBS. The tested package may have dependencies located in those R_LIBS dirs.

**Comment by kforner** [comment]
sorry for asking again: Any reason not to use R_LIBS instead of R_LIBS_USER ? 
This seems the only way to guarantee that the evaluated code will actually use the instrumented package.

**Comment by jimhester** [comment]
Setting `R_LIBS` would break when running examples as I said above. Why can you not just set R_LIBS_USER instead in your package.

Alternatively we can set all of the variables to the same library path, e.g. https://github.com/jimhester/covr/issues/166#issuecomment-212617183.

I was trying to minimize the number of changes needed to the environment, but maybe we have no choice.

**Comment by gaborcsardi** [comment]
Btw. @jimhester it seems that `callr` works well, so if you are interested using it, I can submit a PR. It does the same with the lib environment variables: https://github.com/MangoTheCat/callr/blob/77ef315566ce2ec5300898ac45106b5027309921/R/run.R#L25-L27

The promise is that it will run an arbitrary R function in a separate session, without affecting the current session. By default it sets the lib path to `.libPaths()` in the parent.

It is very hard to get these things right, unfortunately. Only time and testing will tell if this is really the best solution.

**Comment by kforner** [comment]
>Setting R_LIBS would break when running examples as I said above.
ah, I misunderstood what you meant. I thought you were saying that because testInstalledPackage unset R_LIBS it could not work for test tests. 

>Why can you not just set R_LIBS_USER instead in your package.
I do not use R_LIBS in my package, but you can inherit if from your shell config. 
And since it leads to bugs difficult to understand, it's probably good to make it more robust.

>Alternatively we can set all of the variables to the same library path, e.g. #166 (comment).

Or only prepend the covr path to all variables. That's probably the most robust way.

>I was trying to minimize the number of changes needed to the environment, but maybe we have no choice.

I understand, but the impact is probably acceptable
